Cryogenics Overhaul

neuPanda opened the pull request:
adding cryo chems to work after death
balancing cryo chem formulas based on objectives
balancing cryo chems healing based on formulas
making cryo pods more efficient
modifying how the body metabolize chemical mixes
adding a way to split solution reagents evenly
renaming Tramoxadone to Traumoxadone

<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->
added new cryo chem stelloxadone

modified dexo cryo chem to be a more complex solution to fall in line with other cryo chems but also improving its capabilities

modified cryox to not be an omni heal and negate the point of cryo chems but added blood restoration and significantly improved its asphyxiation  healing

modified alox cryo chem to heal cold damage in addition to other burn damages so that omnizine was no longer the only way to heal dead cold damage

all cryo chems now work on the dead

tramoxadone is now traumoxadone

cryo pod is more efficient at administering chemicals

the body can now metabolize multiple meds at once,  still preventing from poison power gaming (consider adding more medical mixes that result in dangerous chemicals like razorium to prevent medical power gaming)
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->

## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->
this is in my effort to help improve medical over all and making cryo a more impactful way of healing to make it actually used in game rather than being a joke concept.  

## How to test
<!-- Describe the way it can be tested -->
- created a cryo setup
- smashed urist with throngler
- ensure that a beaker with stellox, traumox, alox, and dex-ox, and cryox, would all be injected into the body at .25 units each note that this should also be the same even if the units in solution are not balanced
- observe that the body has .5 of each injected into the body and they they are all being metabolized as none are poisons
- change solution to test each chem individually to ensure they heal as expected note that stell heals both poison and radiation.
- tested that cryo heals blood loss but not other damages and after death
- tested that dex-ox heals genetic more effectively and after death 
- tested that alox heals cold
- tested that all of traumox's tests are accurate
- tossed 5 different chems into a beaker
- tested only 3 chems would process at a time randomly chosen
- repeated this test with medicals and narcotics also mixed with the poisons and ensured that everything was injected except poisons, max 3 poisons were injected into the body

## Media
<!-- 
PRs which make in game changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

- [x] I have added screenshots/videos to this PR showcasing its changes in game, **or** this PR does not require an in game showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

:cl:
- add: new cryo chems
- tweek: cryo chems to work after death
- tweek cryo tube more efficient

whatston3 commented on Resources/Locale/en-US/reagents/meta/medicine.ftl:
```diff
@@ -17,7 +17,7 @@ reagent-name-cryoxadone = cryoxadone
 reagent-desc-cryoxadone = Required for the proper function of cryogenics. Heals all standard types of damage, but only works in temperatures under 213K. It can treat and rejuvenate plants when applied in small doses.
```
Add info about cryoxadone working on the dead.
```suggestion
reagent-desc-cryoxadone = Required for the proper function of cryogenics. Heals all standard types of damage, but only works in temperatures under 213K. It can treat and rejuvenate plants when applied in small doses. Works regardless of the patient being alive or dead.
```

whatston3 commented on Content.Server/Body/Components/MetabolizerComponent.cs:
```diff
@@ -55,12 +55,12 @@ public sealed partial class MetabolizerComponent : Component
         /// </summary>
         [DataField("maxReagents")]
-        public int MaxReagentsProcessable = 3;
+        public int MaxPoisonProcessable = 3;
```
Nitpick - this should be plural.
```suggestion
        public int MaxPoisonsProcessable = 3;
```

whatston3 reviewed (changes requested):
Finished code review.  A good few suggestions, most nitpicky (sorry).  Looks interesting!  Testing changes locally.

whatston3 commented on Content.Server/Body/Components/MetabolizerComponent.cs:
```diff
@@ -55,12 +55,12 @@ public sealed partial class MetabolizerComponent : Component
         /// </summary>
         [DataField("maxReagents")]
-        public int MaxReagentsProcessable = 3;
+        public int MaxPoisonsProcessable = 3;
```
I'm not sure if I understand the rationale behind the metabolism change.

As-is, chemicals will eventually metabolize if there are more than can be metabolized at once in the solution - is this to prevent buildup beyond stomach capacity to prevent cryo chem waste with the per-reagent injection?

If anything, species-specific care might be more nuanced with the metabolism as it was (oh, don't put too much in at once, the human/moth might not metabolize it all), and this disincentivizes the 5-in-1 cryo panacea cocktail.

neuPanda replied:
honestly i didn't even want to leave the poison limit, however because people are sweaty power gamers which required this to be added in the first place as to not do 20+ damage per second, i have left the poison restriction behind

dvir001 replied:
Do not change an upstream value like this by name, this is a fucking nightmare to deal with on upstream updates

neuPanda commented on Content.Shared/Chemistry/Components/Solution.cs:
```diff
@@ -606,7 +606,11 @@ public Solution SplitSolutionWithOnly(FixedPoint2 toTake, params string[] includ
+        /// <summary>
+        /// splits the solution taking the specified amount of reagents proportionally to their quantity.
+        /// </summary>
+        /// <param name="toTake">The amount of solution to remove and return</param>
```
will use a different re-wording but this implies that you will only be taking .25 out of the solution, not .25 of each reagent of the solution

whatston3 reviewed (changes requested):
Testing worked alright - stelloxadone is fun to use, gives more use for doxarubixadone, and cryoxadone works on the dead.  All good there.

I was able to create a disgusting slurry of 10+ food ingredients and they all seemed to metabolize at once, I don't quite see what you're trying to do here, though.  Toxins are limited, as written - that works well.

I think necrosol could use a buff as-is, it doesn't seem worth the resources unless you have access to omnizine-producing plants.  Think a reduction to the metabolism rate for that chem might be useful.

neuPanda commented on Resources/Prototypes/_NF/Recipes/Reactions/medicine.yml:
```diff
@@ -8,4 +8,18 @@
+      amount: 2
+  products:
+    Stelloxadone: 5
+    Water: 3
+    Fiber: 2
```
plants have a lot of water and fiber in them.  this was so that rather than saying if i mix a volume of 10 units together i only get 5 units out, i have 10 units and i get out 10 units.  just only 5 of those units were of what you wanted the other 5 are "waste materials" such as the water and the fiber that were in the plants chemicals but have now been removed via reaction.  i could probably be sold on it being something else but if i remember correctly the only thing you get out of galaxy thistle is the stellibinin which means the water and fibers are still in that mixture of plant chemical.  i figure balancing will be happening over time to find a good feal for it as its played

neuPanda commented on Resources/Prototypes/_NF/Reagents/medicine.yml:
```diff
@@ -19,3 +19,28 @@
+            Radiation: -3
+            Cellular: 1
+          groups:
+            Brute: 3
```
the brute 3 is 1 brute, 1 puncture, 1 slash.  which that causes bleeding.  as for internal/external that is not a mechanic i am currently aware of being available to me

neuPanda commented on Resources/Locale/en-US/_NF/reagents/meta/medicine.ftl:
```diff
@@ -1,2 +1,5 @@
+reagent-desc-traumoxadone = A cryogenics chemical. Used to treat severe trauma via regeneration of the damaged tissue. Works regardless of the patient being alive or dead.
+
+reagent-name-stelloxadone = stelloxadone
+reagent-desc-stelloxadone = A cryogenics chemical. Used to agressivly disolve toxins from the body. Works regardless of the patient being alive or dead.
```
the idea is that its removing the toxic buildup but cells are being damaged and destroyed by the reagent at the same time.  the cells are not the target of what is to be damaged, but an unfortunate side effect of the bacteria that are able to dissolve the toxins

neuPanda commented on Content.Shared/Medical/Cryogenics/CryoPodComponent.cs:
```diff
@@ -38,7 +38,14 @@ public sealed partial class CryoPodComponent : Component
-    public float BeakerTransferAmount = 1f;
+    public float BeakerTransferAmount = .25f;
+
+    /// <summary>
+    /// How potent (multiplier) the reagents are when transferred from the beaker to the mob.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite)]
+    [DataField("PotencyAmount")]
+    public float PotencyAmount = 2f;
```
this is a multiplier / factor, not additive. but ill go ahead and re-name it and fiddle with the comment some

whatston3 reviewed (approved):
Tested the most recent changes locally.  Works as before, cryoxadone changes show up in the guidebook chemical entry.  Good stuff.  I've got a [suggested Cryogenics guidebook entry up](https://github.com/whatston3/frontier-station-14/commit/dcf5002b7056799833649a346fd0be518da735eb), feel free to use or change it as you'd like.

I don't like that cryo pods completely avoid overdose with this set of changes (except with Pyrazine), and I think they'll need a nerf, but this seems to behave as you wanted it to.  I'll approve the changes, but I think some rework or balance is needed.  Guess we'll see how cloning fits in with new med?

neuPanda commented:
> I don't like that cryo pods completely avoid overdose with this set of changes (except with Pyrazine), 

this is part of the reason i added side effects to medications this way its a balancing act and needing to measure out what your going to need as to not cause damage while not providing any healing.  ie if you have 0 tox you will still take the brute an genetic from stelloxadone.  I believe more chems need negatives to make them feel more realistic or like in the case of the brute chems have dangerous mixtures if combine with each other.  but without some feedback from people playing and indicating how things feel its hard to balance

>Another note - should this potency be applied to meds other than cryo-specific ones? I'd argue no - if it was, I can't see a reason to skip using cryo pods on a hospital ship, even if only to save on meds.

part of this is a limitation of the code i would have to go in and create a system that can be used to group different med types and if not done right that can get really messy really fast.
another part being that cryopods are advance machines that have the ability to get the meds where they are needed with less wasted material seeping into parts of the body not needing the treatment



in the future something i think would be neat would be to be able to have the cryo pod be comprised of multiple components rather than glass and wire.  then as components are upgraded you can increase the potency(manipulator?), how many chems it can inject at a time (matter bin?), so base, 
2 chems, at a 1.15 multiplier, 
3 chems at a 1.40 multiplier, 
4 chems at a 1.9 multiplier, 
5 chems at a 2.5 multiplier

dvir001 reviewed (changes requested):
You also didn't add a note ``// Frontier`` to any of the added lines making it hard to track down on upstream updates and in yml and c#

dvir001 commented:
Your Changelog are not how the Changelog used, this will not showup in game at all but just bug out.

dvir001 commented:
> > Your Changelog are not how the Changelog used, this will not showup in game at all but just bug out.
> 
> please utilize the constructive part of criticism do not tell people "this is wrong" tell them "this is wrong, here is how to improve it"
> 
> what is it i need to change on my change log because all you have provided is that it is wrong, now how to improve it

```
**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

<!--
Make sure to take this Changelog template out of the comment block in order for it to show up.
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->
```

Use the format to make it show in game

neuPanda commented:
> ```
> <!--
> Make sure to take this Changelog template out of the comment block in order for it to show up.
> :cl:
> - add: Added fun!
> - remove: Removed fun!
> - tweak: Changed fun!
> - fix: Fixed fun!
> -->
> ```

> Use the format to make it show in game, try to keep it into 3 changelog lines top as to not fill the changelog too much

ah, thank you.  that has been updated to keep it short and sweet.  i will double check that i didn't miss that step in my other one.  thank you for the clarification on what i had missed.

whatston3 commented:
I've given this branch another look - I believe I have a suggestions branch that I'm happy with in terms of code.

The branch itself is at [whatston3:2024-06-12-CryogenicsOverhaul-suggestions](https://github.com/whatston3/frontier-station-14/tree/2024-06-12-CryogenicsOverhaul-suggestions).  This has a merge with the current head of master, [a commit with all of the cleanup for this PR](https://github.com/new-frontiers-14/frontier-station-14/commit/5fcc2d1676214eab334b83d2efb5b8fe2fd7b852), and a series of guidebook changes and cleanup.

This suggestion branch:
* Removes maxPoisonsApplied - species that metabolize more things should have different effects compared to each other.  Now it is a negative rather than having pros and cons, but there should still be differences between a slime and a dwarf.  Since none of the species had a definition for maxPoisonsApplied, this would've led to the same default behaviour across all species.
*  Adds more descriptive Frontier/End Frontier sections to the comments, in a more consistent style with the rest of the codebase.
* Simplifies the SplitSolutionReagentsEvenly function, making it similar to the other SplitSolution functions around it.
* Alters the localized descriptions of cryogenic medications where appropriate.  The writeups might need some work, but it's more accurate than before.

This is my best attempt at getting this revised without changing behaviour.

That said, behaviourally, I still think that cryo pods are _insanely buffed_ in this change - it is now impossible to overdose with any medication in the cryo pod (save Pyrazine, or with Arachnids?), and trivializes healing live patients.  You can create a cocktail of all of the basic drugs, put that into a keg, pour out a beaker with it as needed, and only use that to heal anyone that comes in without a thought.

I like the new drugs!  I like the idea of buffing cryo pods!  This is a good feature, I want this in.  However, the way drugs are filtered is odd and avoids most of the downsides of metabolism as-is.  If you fix that, I'm happy.  More than willing to hash this out, just let me know.

Cheackraze commented:
conceptually I love all the things happening here, just waiting for the code to be ready steady. personally, medical and chemistry are the areas of the game I have spent the least amount of time doing, since I find the current systems to be so flat and one dimensional, but I hope this is a good step in the right direction before we get any kind of new med
